mne/source_estimate.py

```diff
@@ -1098,6 +1097,74 @@ def transform(self, func, idx=None, tmin=None, tmax=None, copy=False):
 
         return stcs
 
+    @verbose
+    def estimate_snr(self, info, fwd, cov, verbose=None):
+        r"""Compute time-varying SNR in the source space.
+
+        This function should only be used with source estimates with units
+        nanoAmperes (i.e., MNE-like solutions, *not* dSPM or sLORETA).
+
+        Parameters
+        ----------
+        info : instance of io.meas_info.Info
+            The measurement info.
+        fwd : instance of Forward
+            The forward solution used to create the source estimate.
+        cov : instance of Covariance
+            The noise covariance used to estimate the resting cortical
+            activations. Should be an evoked covariance, not empty room.
+
+        Returns
+        -------
+        snr_stc : instance of SourceEstimate
+            The source estimate with the SNR computed.
+
+        Notes
+        -----
+        We define the SNR in decibels for each source location at each
+        time point as:
+
+        .. math::
+
+            {\rm SNR} = 10\log_10[\frac{a^2}{N}\sum_k\frac{b_k^2}{s_k^2}]
+
+        where :math:`\\b_k` is the signal on sensor :math:`k` provided by the
+        forward model for a source with unit amplitude, :math:`a` is the
+        source amplitude, :math:`N` is the number of sensors, and
+        :math:`s_k^2` is the noise variance on sensor :math:`k`.
+        
+        Reference:
+        Goldenholz, D. M., Ahlfors, S. P., Hämäläinen, M. S., Sharon, D.,
+        Ishitobi, M., Vaina, L. M., & Stufflebeam, S. M. (2009).
+        Mapping the Signal-To-Noise-Ratios of Cortical Sources in
+        Magnetoencephalography and Electroencephalography.
+        Human Brain Mapping, 30(4), 1077–1086. doi:10.1002/hbm.20571
+        """
+        from .forward import convert_forward_solution
+        from .minimum_norm.inverse import _prepare_forward
+        
+        M = np.shape(self.data)[0]
+        T = np.shape(self.data)[1]
+        snr_stc = SourceEstimate(np.zeros((M, T)), self.vertices, self.tmin, 
+                                 self.tstep, self.subject)
+        
+        fwd = convert_forward_solution(fwd, surf_ori=True, force_fixed=True)
+        
+        # G is gain matrix [ch x src], cov is noise covariance [ch x ch]
+        _, _, G, _, _, _, _, cov, _ = _prepare_forward(fwd, info, cov,
+                                fixed=True, loose=0, rank=None, pca=False,
+                                use_cps=True, exp=None, limit_depth_chs=False,
+                                combine_xyz='fro', allow_fixed_depth=True,
+                                limit=None)
+
+        n_channels = cov['dim'] # number of sensors/channels    
+        b_k2 = (G * G).T
+        s_k2 = np.diag(cov['data'])                
+        scaling = (1 / n_channels) * np.sum(b_k2 / s_k2, axis=1)[:, np.newaxis]  
+        snr_stc.data = 10 * np.log10((self.data * self.data) * scaling)
```

Comment:
instantiate the `snr_stc` object here so you avoid initializing another object with zeros (thus twice the memory)

Reply:
all set